### ISSUE TITLE
Instantiation Context Recording for Deferred Operations

### ISSUE THREAD

**Issue by rcaudy** [deephaven/deephaven-core#2496]
As a user, when I invoke an operation that may be deferred, I want the `QueryScope`, `QueryLibrary`, and `CompilerTools.Context` that were active at invocation time to apply when the operation is ultimately applied. Otherwise, it can be difficult to ensure correct operation if this context is mutated between my invocation and eventual instantiation.

This has a few aspects:
1. Deferred views and filters applied to `UncoalescedTable`
2. Update propagation for `Table.update()` and `Table.select()` with a `Table` as input
3. Update propagation for `PartitionedTable.transform()` and `PartitionedTable.partitionedTransform()`, as well as operations invoked via `PartitionedTable.Proxy` _(These are refinements of case 2, but we may be able to automatically narrow the scope of the context we record to avoid unnecessarily preserving `Object`s past their intended lifetime.)_
4. User listeners that may apply operations to `Table`s on notification

This is important to deliver for multiple reasons:
1. Users may choose to mutate the `QueryScope` or other global context after invoking deferred operations, incorrectly assuming that this is already guaranteed to be safe
2. Users may choose to install a different `QueryScope`, `QueryLibrary`, or `CompilerTools.Context` with the intention of isolating invocations from the global context, and we should fully honor this intention
3. Python modules should be able to correctly expose their own globals to the query engine without polluting the enclosing modules globals
4. Isolation is required if we ever want to support multiple script sessions within the same worker

**Comment by rcaudy** [comment]
Summarizing our intended design a bit after meeting with @cpwright and @nbauernfeind :

# Execution Context
1. We will introduce an `ExecutionContext` that immutably bundles the global state we're managing, as in the current WIP PR (#2539).
2. This will be installed for parallel instantiation (used in `update` and `select`), and recorded by `Listener` implementations to be reinstalled during `Notification` delivery (requires special consideration in parallel `update` or `select`).
3. If the instantiation context is the "default" one, `Listener` implementations will actually record a "poisoned" context that is empty and will throw exceptions on attempted use. That is, formulas and conditions are unable to compile code or reference the query scope at *invocation* time unless the user has taken special steps. This eliminates the expectation that we will always re-install the instantiation context when it's usually unnecessary, as well as the need to automatically record the entire `QueryScope` and thereby prevent GC for all scope params.
4. Defining "default": By default, there is no `ExecutionContext` installed, operations can reference the context `ThreadLocal`s as normal, and the poisoned execution context will be recorded so that notifications cannot reference the `ThreadLocal`s.
5. `ExecutionContext` should have a nice builder pattern for constructing limited scopes or inherited limited views of the `ThreadLocal`s (e.g. "take everything from `QueryLibrary` and `CompilerTools.Context`, but only take a named subset of params from `QueryScope`).
6. `ExecutionContext` should have a `try-with-resources` pattern for scoping initialization or notification processing.

# Partitioned Tables:
1. We will update `PartitionedTableProxyImpl` to be sure that each set of method overloads delegates to a single core overload per operation.
2. We will ensure that `Selectable` and `Filter` objects used in proxied operations are fully instantiated (thereby recording everything they need from the current contexts) before applying them (that is, before calling `transform` or `partitionedTransform`).
3. We will either overload `transform` and `partitionedTransform` to explicitly accept an `ExecutionContext`, or we will document the possible need for one in the JavaDoc. This gives us a mechanism to ensure that users are aware that if they are doing complex transforms (for example, in order to chain table operations), they may need to use an `ExecutionContext`.

# Deferred Operations (`DeferredViewTable` and `RedefinableTable`):
We will either record `ExecutionContext` information with every deferred `SelectColumn` or `SelectFilter` and ensure that they are instantiated with the appropriate context installed, or we will attempt to force initialization at time of deferrment.
